week4/path/vitcra/Main.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE RecordWildCards #-}
+module Main where
+
+import Data.Set (Set)
+import qualified Data.Set as S
+import Data.Array
+import Data.Map (Map)
+import qualified Data.Map as M
+import Data.List
+import Data.Maybe
+import Control.Monad
+import Data.List.Split
+
+type Index = (Int, Int)
+
+data Infinite a = Infinity | Finite a deriving (Show)
+
+data ProblemState = ProblemState
+  { dist :: Array Index (Infinite Int)
+  , edges :: Array Index Char
+  , unvisited :: Set (Infinite Int, Index)
+  , prev :: Map Index Index } deriving (Show)
+
+instance Eq a => Eq (Infinite a) where
+  Infinity == Infinity = True
+  (Finite a) == (Finite b) = a == b
+  _ == _ = False
+
+instance Ord a => Ord (Infinite a) where
```

Comment:
This instance could be automatically derived if you changed the order in your declaration:

```
data Infinite a = Finite a | Infinity
```